### PR TITLE
Update warning setup_context

### PR DESCRIPTION
**Pull request by oow214**
Update warning module https://github.com/RustPython/RustPython/issues/4013
- [x] is_internal_frame
- [x] next_external_frame
- [x] setup_context

### CODE REVIEW

File: vm/src/warn.rs [hunk 1]
```diff
@@ -72,32 +72,39 @@ fn setup_context(
...
+                while let Some(frame) = f.clone() {
+                    stack_level -= 1;
+                    if stack_level > 0 {
+                        break;
+                    }
+                    f = frame.next_external_frame(vm);
+                }
+            }
+        }
+    }
```
Comment (Snowapril) [inline vm/src/warn.rs:104]:
```suggestion
    if stack_level <= 0 || f.clone().map(|frame| frame.is_internal_frame()).unwrap_or_default() {
        loop {
            stack_level -= 1;
            if stack_level <= 0 || f.is_none() {
                break;
            }
            f = f.unwrap().f_back(vm);
        }
    } else {
        loop {
            stack_level -= 1;
            if stack_level <= 0 || f.is_none() {
                break;
            }
            f = f.unwrap().next_external_frame(vm);
        }
    }
```

Comment (youknowone) [reply vm/src/warn.rs]: I don't support `f.is_none()` and `f.unwrap()` combination when it doesn't give a significant advantage. It looks ok for now, but that works only when human carefully keep the pair in the correct order. We can write the same logic with `if let` matches which prevent mistakes from future logic changes.

File: vm/src/warn.rs [hunk 2]
```diff
@@ -72,32 +72,34 @@ fn setup_context(
...
+        || f.clone()
+            .map(|frame| frame.is_internal_frame())
+            .unwrap_or_default()
+    {
+        loop {
+            stack_level -= 1;
+            if stack_level <= 0 || f.is_none() {
+                break;
+            }
+            f = f.unwrap().f_back(vm);
```
Comment (youknowone) [inline vm/src/warn.rs:89]:
```suggestion
            if stack_level <= 0 {
                break;
            }
            if let Some(f) = f {
                f = f.f_back(vm);
            } else {
                break;
            }
```

File: vm/src/frame.rs [hunk 3]
```diff
@@ -264,6 +264,23 @@ impl FrameRef {
...
+        filename.as_str().contains("importlib") && filename.as_str().contains("_bootstrap")
+    }
+
+    pub fn next_external_frame(&self, vm: &VirtualMachine) -> Option<FrameRef> {
+        self.clone().f_back(vm).map(|back| {
+            if back.is_internal_frame() {
+                back
+            } else {
+                back.f_back(vm).unwrap()
+            }
```
Comment (youknowone) [inline vm/src/frame.rs:281]: The function name and its behavior looks confusing.
The name is `next_external_frame` but the condition is `back.is_internal_frame()` than return the value.
If this is a correct implementation, we'd better to add comment how this is working.

File: vm/src/warn.rs [hunk 4]
```diff
@@ -72,32 +72,41 @@ fn setup_context(
...
-    //         while (--stack_level > 0 && f != NULL) {
-    //             f = next_external_frame(f);
-    //         }
-    //     }
+    // Stack level comparisons to Python code is off by one as there is no
+    // warnings-related stack level to avoid.
+    if stack_level <= 0
+        || f.clone()
+            .map(|frame| frame.is_internal_frame())
+            .unwrap_or_default()
```
Comment (youknowone) [inline vm/src/warn.rs:82]: maybe redundant clone?
```suggestion
        || f.map_or(false, |frame| frame.is_internal_frame())
```

File: vm/src/warn.rs [hunk 5]
```diff
@@ -72,32 +72,41 @@ fn setup_context(
...
+    {
+        loop {
+            stack_level -= 1;
+            if stack_level <= 0 || f.is_none() {
+                break;
+            }
```
Comment (youknowone) [inline vm/src/warn.rs:88]: redundant check with line 89 and 92

File: vm/src/warn.rs [hunk 6]
```diff
@@ -72,32 +72,41 @@ fn setup_context(
...
+            if let Some(tmp) = f {
+                f = tmp.f_back(vm);
+            } else {
+                break;
+            }
+        }
+    } else {
+        loop {
+            stack_level -= 1;
+            if stack_level <= 0 || f.is_none() {
```
Comment (youknowone) [inline vm/src/warn.rs:101]: the `is_none()` and `unwrap()` pair here also can be eliminated by using line 89 and 92 patterns

**Review by youknowone — COMMENTED** [review]
looks good in general. I have a question about next_external_frame and a few more small change requests.

File: vm/src/frame.rs [hunk 7]
```diff
@@ -264,6 +264,29 @@ impl FrameRef {
...
+    // If the back frame is an internal frame, it is judged as a new external frame and the loop is stopped.
+    pub fn next_external_frame(&self, vm: &VirtualMachine) -> Option<FrameRef> {
+        self.clone().f_back(vm).map(|mut back| loop {
+            back = if let Some(back) = back.to_owned().f_back(vm) {
+                back
+            } else {
+                break back;
+            };
+
+            if back.is_internal_frame() {
```
Comment (youknowone) [inline vm/src/frame.rs:285]: maybe
```suggestion
            if !back.is_internal_frame() {
```

File: vm/src/warn.rs [hunk 8]
```diff
@@ -72,32 +72,38 @@ fn setup_context(
...
-    //         while (--stack_level > 0 && f != NULL) {
-    //             f = next_external_frame(f);
-    //         }
-    //     }
+    // Stack level comparisons to Python code is off by one as there is no
+    // warnings-related stack level to avoid.
+    if stack_level <= 0 || f.as_ref().map_or(false, |frame| frame.is_internal_frame()) {
+        loop {
+            stack_level -= 1;
+            if stack_level <= 0 || f.is_none() {
```
Comment (youknowone) [inline vm/src/warn.rs:82]:
```suggestion
            if stack_level <= 0 {
```

**Review by youknowone — CHANGES_REQUESTED** [review]

**Review by youknowone — APPROVED** [review]
Looks good, thank you!

**Pull request merged**